src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs

```diff
@@ -133,58 +134,96 @@ private Task<ChecksumCollection> BuildAsync(IReadOnlyList<AnalyzerReference> key
             return CreateChecksumCollectionsAsync(documentStates, kind, snapshotBuilder.BuildAsync, cancellationToken);
         }
 
-        private Task<ChecksumCollection> CreateChecksumCollectionsAsync(
+        private async Task<ChecksumCollection> CreateChecksumCollectionsAsync<TValue, TChecksumObject>(
+            IEnumerable<TValue> items, string kind, Func<TValue, CancellationToken, Task<TChecksumObject>> buildAsync, CancellationToken cancellationToken) where TChecksumObject : ChecksumObject
+        {
+            using (var pooledObject = Creator.CreateList<Task<TChecksumObject>>())
+            {
+                // create asyn checksums concurrently
+                var tasks = pooledObject.Object;
+                foreach (var item in items)
+                {
+                    tasks.Add(buildAsync(item, cancellationToken));
+                }
+
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+
+                var checksums = new Checksum[tasks.Count];
+
+                for (var i = 0; i < tasks.Count; i++)
+                {
+                    var task = tasks[i];
+
+                    // task is already completed. so wait will return right away as long as there is no exception.
+                    // if there is exception, then it will throw it. if it is cancellation exception, it will make sure
+                    // it throws cancellation exception rather than aggregated exception.
+                    task.Wait(cancellationToken);
```

Comment:
I have no concern with the comments.  My concern is with calling task.Wait.  That is innapropriate and we should simply 'await' here to get the right behavior.  By calling 'Wait' we now have to deal with audting this code every time we go through looking for blocking calls in async methods.  .Wait in an async context is traditionally a large red flag that something is wrong.  That auditing will waste time and make this code different for no valuable reason. 

The appropriate thing to do here is to just 'await' the task like we would in pretty much all other code.  

Now, personally, that code is clear enough that it should need no comment.  But i have no problem with the code calling out why it does ```await task.ConfigureAwait(false)``` rather than just ```task.Result```.